internal/graph/membershipset.go

```diff
@@ -0,0 +1,254 @@
+	for resourceID, details := range resultsMap {
+		existing, ok := ms.membersByID[resourceID]
+		if !ok {
+			continue
+		}
+
+		ms.membersByID[resourceID] = caveatAnd(existing, details.Expression)
+	}
+
+	hasDeterminedMember := false
+	for resourceID, expr := range ms.membersByID {
+		if _, ok := resultsMap[resourceID]; !ok {
+			delete(ms.membersByID, resourceID)
+			continue
+		}
+
+		if expr == nil {
+			hasDeterminedMember = true
+		}
+	}
+
+	ms.hasDeterminedMember = hasDeterminedMember
+}
+
+// Subtract subtracts the results found in the given map with the members of this set.
+// The changes are made in-place.
+func (ms *MembershipSet) Subtract(resultsMap CheckResultsMap) {
+	for resourceID, details := range resultsMap {
+		existing, ok := ms.membersByID[resourceID]
+		if !ok {
+			continue
+		}
+
+		// If the incoming result has no caveat, then this removal is absolute.
+		if details.Expression == nil {
+			delete(ms.membersByID, resourceID)
+			continue
+		}
+
+		// Otherwise, the caveat expression gets combined with an intersection of the inversion
+		// of the expression.
+		ms.membersByID[resourceID] = caveatSub(existing, details.Expression)
+	}
+
+	ms.hasDeterminedMember = false
+	for _, expr := range ms.membersByID {
+		if expr == nil {
+			ms.hasDeterminedMember = true
+			break
+		}
+	}
+}
+
+// IsEmpty returns true if the set is empty.
+func (ms *MembershipSet) IsEmpty() bool {
+	return len(ms.membersByID) == 0
+}
+
+// HasDeterminedMember returns whether there exists at least one non-caveated member of the set.
+func (ms *MembershipSet) HasDeterminedMember() bool {
+	return ms.hasDeterminedMember
+}
+
+// AsCheckResultsMap converts the membership set back into a CheckResultsMap for placement into
+// a DispatchCheckResult.
+func (ms *MembershipSet) AsCheckResultsMap() CheckResultsMap {
+	resultsMap := make(CheckResultsMap, len(ms.membersByID))
+	for resourceID, caveat := range ms.membersByID {
+		membership := v1.DispatchCheckResponse_MEMBER
+		if caveat != nil {
+			membership = v1.DispatchCheckResponse_CAVEATED_MEMBER
+		}
+
+		resultsMap[resourceID] = &v1.DispatchCheckResponse_ResourceCheckResult{
+			Membership: membership,
+			Expression: caveat,
+		}
+	}
+
+	return resultsMap
+}
+
+func wrapCaveat(caveat *core.ContextualizedCaveat) *v1.CaveatExpression {
+	if caveat == nil {
+		return nil
+	}
+
+	return &v1.CaveatExpression{
+		OperationOrCaveat: &v1.CaveatExpression_Caveat{
+			Caveat: caveat,
+		},
+	}
+}
+
+func caveatOr(first *v1.CaveatExpression, second *v1.CaveatExpression) *v1.CaveatExpression {
+	if first == nil {
+		return second
+	}
+
+	if second == nil {
+		return first
+	}
+
+	return &v1.CaveatExpression{
+		OperationOrCaveat: &v1.CaveatExpression_Operation{
+			Operation: &v1.CaveatOperation{
+				Op:       v1.CaveatOperation_OR,
+				Children: []*v1.CaveatExpression{first, second},
+			},
+		},
+	}
+}
+
+func caveatAnd(first *v1.CaveatExpression, second *v1.CaveatExpression) *v1.CaveatExpression {
```

Comment:
I guess that while doing those logical operations at CEL level is plausible, it would make things more complicated to test as we need to introspect the resulting CEL. Another consideration is we would need to load the caveat from database earlier, we may be able to optimize further if we batch load those once we need to resolve the chain of caveats. But if we need to load the tuples, we may as well join it with the caveats tuple so that wouldn't be a concern.

Did you consider handle the logical operations at CEL level, and if so, why not?

Reply:
Its a tradeoff: loading the caveats in the relationship loader and combining them via CEL means we might get some early termination (depending on the data already found for the relationship), but it does mean carrying around a lot more data if the expressions are larger. It also makes testing of the MembershipSet much harder and binds us directly to CEL. We may very well change over to it, but this approach seemed simpler for the initial pass.